client/engine/chainservice/eth_chainservice.go

```diff
@@ -116,35 +155,131 @@ func (ecs *EthChainService) SendTransaction(tx protocols.ChainTransaction) error
 		return fmt.Errorf("unexpected transaction type %T", tx)
 	}
 }
-func (ecs *EthChainService) subcribeToEvents() error {
 
-	go ecs.listenForLogEvents()
-	return nil
+// fatalError is called to output the error and then panic, killing the chain service.
+// If prints out the error to STDOUT, the logger and then exits the program.
+func (ecs *EthChainService) fatalError(err error) {
+	ecs.fatalF("FATAL ERROR\n%+v", err)
+}
+
+// fatalF is called to output a message and then panic, killing the chain service.
+// It accepts a format string and arguments, as per fmt.Printf.
+// If prints out the error to STDOUT, the logger and then exits the program.
+func (ecs *EthChainService) fatalF(format string, v ...any) {
+
+	// Print to STDOUT in case we're using a noop logger
+	fmt.Println(fmt.Errorf(format, v...))
+	// FatalF prints to the logger then calls exit(1)
+	ecs.logger.Fatalf(format, v...)
+
+	// Manually panic in case we're using a logger that doesn't call exit(1)
+	panic(fmt.Errorf(format, v...))
+
 }
 
-func (ecs *EthChainService) listenForLogEvents() {
-	// Subsribe to Adjudicator events
+// dispatchChainEvents takes in a collection of event logs from the chain
+// and dispatches events to the out channel
+func (ecs *EthChainService) dispatchChainEvents(logs []ethTypes.Log) {
+	for _, l := range logs {
+		switch l.Topics[0] {
+		case depositedTopic:
+			nad, err := ecs.na.ParseDeposited(l)
+			if err != nil {
+				ecs.fatalF("error in ParseDeposited: %v", err)
+			}
+
+			event := NewDepositedEvent(nad.Destination, l.BlockNumber, nad.Asset, nad.AmountDeposited, nad.DestinationHoldings)
+			ecs.out <- event
+		case allocationUpdatedTopic:
+			au, err := ecs.na.ParseAllocationUpdated(l)
+			if err != nil {
+				ecs.fatalF("error in ParseAllocationUpdated: %v", err)
+
+			}
+
+			tx, pending, err := ecs.chain.TransactionByHash(context.Background(), l.TxHash)
+			if pending {
+				ecs.fatalF("Expected transaction to be part of the chain, but the transaction is pending")
+
+			}
+			var assetAddress types.Address
+			var amount *big.Int
+			switch {
+
+			// TODO: Workaround for https://github.com/filecoin-project/ref-fvm/issues/1158
+			case err != nil && strings.Contains(err.Error(), FEVM_PARSE_ERROR):
+				ecs.logger.Printf("WARNING: Cannot parse transaction, assuming asset address of 0x00...")
+				fmt.Printf("WARNING: Cannot parse transaction, assuming asset address of 0x00...")
```

Comment:
> Or, we write our own parser which extracts only those fields we actually care about? This would still technically be a workaround, but pretty close to something which could just serve as prod code.

Unfortunately, I'm not sure how easy it is to access the underlying data. The deserialization error occurs in [go-ethereum](https://github.com/ethereum/go-ethereum/blob/c2e0abce2eedc1ba2a1b32c46fd07ef18a25354a/rpc/client.go#L348) which means we just get a nil transaction back.

Reply:
Right. I guess if we took on responsibility for crafting the RPC call etc we could do it... but maybe not worth that effort.

Reply:
> Right. I guess if we took on responsibility for crafting the RPC call etc we could do it... but maybe not worth that effort.

Maybe I'm overly optimistic but I'm hoping that the underlying issue with be fixed soon enough that it makes sense to wait for that.

Reply:
Ok I've managed to move the workaround logic test in https://github.com/statechannels/go-nitro/pull/993/commits/f264387ccd262b8e0d165d7943de09df3404ce54. I've created a `fevmWorkoundChain` struct that has a custom implementation of `TransactionByHash` that returns a dummy transaction if it runs into the error.

Reply:
Nice!